### PR TITLE
[z3] Update to v4.8.8

### PR DESCRIPTION
**Pull request by ahumenberger**
https://github.com/Z3Prover/z3/issues/4183

### CODE REVIEW

File: Z/z3/build_tarballs.jl [hunk 1]
```diff
@@ -3,23 +3,36 @@
...
+    ArchiveSource("https://julialang-s3.julialang.org/bin/linux/armv7l/1.3/julia-1.3.1-linux-armv7l.tar.gz", "965c8fab2214f8ce1b3d449d088561a6de61be42543b48c3bbadaed5b02bf824"; unpack_target="julia-arm-linux-gnueabihf"),
+    ArchiveSource("https://julialang-s3.julialang.org/bin/linux/x64/1.3/julia-1.3.1-linux-x86_64.tar.gz", "faa707c8343780a6fe5eaf13490355e8190acf8e2c189b9e7ecbddb0fa2643ad"; unpack_target="julia-x86_64-linux-gnu"),
+    ArchiveSource("https://github.com/Gnimuc/JuliaBuilder/releases/download/v1.3.0/julia-1.3.0-x86_64-apple-darwin14.tar.gz", "f2e5359f03314656c06e2a0a28a497f62e78f027dbe7f5155a5710b4914439b1"; unpack_target="julia-x86_64-apple-darwin14"),
+    ArchiveSource("https://github.com/Gnimuc/JuliaBuilder/releases/download/v1.3.0/julia-1.3.0-x86_64-w64-mingw32.tar.gz", "c7b2db68156150d0e882e98e39269301d7bf56660f4fc2e38ed2734a7a8d1551"; unpack_target="julia-x86_64-w64-mingw32"),
 ]
 
 # Bash recipe for building across all platforms
 script = raw"""
+case $target in
+  x86_64-linux-gnu)
```
Comment (giordano) [inline Z/z3/build_tarballs.jl:20]: Something like
```suggestion
  *-linux-*)
```
?

**Comment by ahumenberger** [comment]
Is it possible to provide platform-specific products (or optional products)? Z3's Julia bindings depend on libcxxwrap which is not available for all `supported_platforms()`, that is, building `libz3jl` is only possible for certain platforms. Other products like `libz3` and `z3` can be built on all platforms.

**Comment by giordano** [comment]
> Is it possible to provide platform-specific set of products (or optional products)?

Yes, it's possible but it's a bit convoluted.  See the builder for [`Qt_jll`](https://github.com/JuliaPackaging/Yggdrasil/blob/b1b143583e30fc1063e2a071031201159e175e73/Q/Qt/build_tarballs.jl#L153-L160) as an example.  The trick consists on calling th `build_tarballs()` function multiple times with different `platforms` and `products` argument.

File: Z/z3/build_tarballs.jl [hunk 2]
```diff
@@ -3,42 +3,84 @@
...
 # Build the tarballs, and possibly a `build.jl` as well.
-build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; preferred_gcc_version=v"8")
+include("../../fancy_toys.jl")
+
+if any(should_build_platform.(triplet.(platforms_libcxxwrap)))
+    build_tarballs(ARGS, name, version, sources, script, platforms_libcxxwrap, products_libcxxwrap, dependencies_libcxxwrap; preferred_gcc_version=v"8")
+end
+if any(should_build_platform.(triplet.(platforms)))
+    build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; preferred_gcc_version=v"8")
+end
```
Comment (giordano) [inline Z/z3/build_tarballs.jl:86]: Sorry, I just realised that the `Qt_jll` wasn't the best example, we also need to make sure that `--register` is passed only to the last `build_tarballs` calls.  See [this other example](https://github.com/JuliaPackaging/Yggdrasil/blob/c3e3c4a96c723306b4da23fc6d05f12995b21ed8/G/Git/build_tarballs.jl#L76-L93).  The important bit is to pass `non_reg_ARGS` in the first call and the full `ARGS` in the second and last one.

File: Z/z3/build_tarballs.jl [hunk 3]
```diff
@@ -3,42 +3,86 @@
...
     ExecutableProduct("z3", :z3)
 ]
 
 # Dependencies that must be installed before this package can be built
+dependencies_libcxxwrap = Dependency[
+    Dependency("libcxxwrap_julia_jll")
+]
+
 dependencies = Dependency[
 ]
```
Comment (giordano) [inline Z/z3/build_tarballs.jl:76]: Sorry, I spot small issues every time :grimacing:

We can't really have platform-dependent dependencies, the generated JLL package will be only one:
```suggestion
dependencies = [
    Dependency("libcxxwrap_julia_jll")
]
```
You'll also need to use `dependencies` everywhere below.

Comment (ahumenberger) [reply Z/z3/build_tarballs.jl:76]: And I learn something new every time 😉

**Pull request merged**